llvm/include/llvm/Support/Progress.h

```diff
@@ -0,0 +1,256 @@
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/RWMutex.h"
+
+namespace llvm {
+
+class Progress;
+class TaskStack;
+
+/// Track advancement of a task
+///
+/// A task can have a name and optionally a set of maximum steps.
+class Task {
+protected:
+  TaskStack &Stack;
+  int64_t StepIndex = -1;
+  std::optional<int64_t> TotalSteps;
+
+private:
+  std::string TaskName;
+  std::string StepName;
+  bool SubtaskCreated = false;
+  bool Completed = false;
+  bool CurrentStepHasSingleSubtask = false;
+
+public:
+  Task(std::optional<int64_t> TotalSteps, const llvm::Twine &TaskName = "");
+
+  Task(TaskStack &Stack, std::optional<int64_t> TotalSteps,
+       const llvm::Twine &TaskName = "");
+
+  Task(const Task &) = delete;
+  Task(Task &&) = delete;
+
+  ~Task();
+
+public:
+  llvm::StringRef name() const { return TaskName; }
+  llvm::StringRef stepName() const { return StepName; }
+  int64_t stepIndex() const {
+    return std::max(static_cast<int64_t>(0), StepIndex);
+  }
+  std::optional<int64_t> totalSteps() const { return TotalSteps; }
+  bool subtaskCreated() const { return SubtaskCreated; }
+  bool currentStepHasSingleSubtask() const {
+    return CurrentStepHasSingleSubtask;
+  }
+  bool completed() const { return Completed; }
+  size_t index() const;
+
+  TaskStack &stack() { return Stack; }
+  const TaskStack &stack() const { return Stack; }
+
+public:
+  /// \param NewStepName the name of the new step starting after this advance
+  ///        invocation.
+  /// \param SingleSubtask whether we can expect 0 or 1 subtask being created
+  ///        during this step. If true, will assert in case multiple subtasks
+  ///        are create during this step.
+  void advance(const llvm::Twine &NewStepName = "", bool SingleSubtask = false);
+
+  void setSubtaskCreated();
+
+  void complete();
+
+};
+
+/// Represents the set of tasks started but not completed in an execution thread
+class TaskStack {
+private:
+  Progress &Parent;
+  unsigned SuspendRequests = 0;
+
+public:
+  llvm::SmallVector<Task *, 4> Tasks;
+
+public:
+  TaskStack(Progress &Parent) : Parent(Parent) {}
+  TaskStack(TaskStack &&) = default;
+
+  TaskStack(const TaskStack &) = delete;
+  TaskStack &operator=(const TaskStack &) = delete;
+  TaskStack &operator=(TaskStack &&) = delete;
+
+public:
+  void registerTask(Task *T);
+  void unregisterTask(Task *T);
+  void advanceTask(Task *T, llvm::StringRef PreviousStepName);
+
+public:
+  bool isSuspended() const { return SuspendRequests > 0; }
+  void resumeTracking() {
+    assert(isSuspended());
+    --SuspendRequests;
+  }
+
+  void suspendTracking() { ++SuspendRequests; }
+};
+
+/// Class to monitor progress of a Progress instance
+class ProgressListener {
+public:
+  virtual ~ProgressListener() = default;
+
+  /// Invoked to notify the creation of a new task
+  ///
+  /// \note \p T will always be the last task of its stack.
+  ///
+  /// \note This method could be called by multiple threads in parallel.
```

Comment:
Advertising this on the API is tempting, because it's easy and it delegates resposibility to implementors. However, this has already proven to be error-prone in `revng`, even the implementor it's you who designed the API.

I'd suggest to put a mutex in `ProgressListener` too. Then you could have:

```c++
class ProgressListener {
/// ...

private:

void handleNewTaskThreadSafely(const Task *T) {
  // acquire mutex
  handleNewTask(T);
  // release mutex
}

public:
// Implementors implement this, and they don't need to care about threads
// calling this concurrently, because it's guarded by the previous line.
virtual void handleNewTask(const Task *T)  = 0;

};
```

Then `Progress::handleNewTask` should just call `handleNewTaskThreadSafely` and that's it.

Reply:
Then something similar needs to be done for the other 2 pure virtual methods, but the idea is the same.

Reply:
Eh no, I don't want to do this.
I want to push the responsibility on the listeners so they can be more efficient.

Consider that accessing `Task` and `TaskStack` do not require any synchronization, they are thread specific.
What needs synchronization is emitting the output, but that's a listener's responsibility to decided whether to have per-thread buffers and commit every now and then, or commit all the time, or write to different files...

Reply:
Ok I just wanted to make sure that you were aware of this, that you're willing to push responsibility onto implementers in exchange for potential additional performance.

But there are still a few details that I want to be extra sure of. You're saying in the comments that `handleNewTask` (and friends) on listeners should be callable concurrently by many threads. Given that there's only a single call to that in `Progress::handleNewTask` (and friends), this means that you expect `Progress::handleNewTask` (and friends) to be called concurrently. So what happens if 2 different threads call the `handleNewTask` method of the same listener concurrently? That would cause `handleNewTask` to be called concurrently on the same Listener. This means that all the listener need to guard their internal state. But what I want to stress is that it's not enough for `Task` and `TaskStruct` to be thread-specific, you need all the listeners to be designed for concurrent access to their internal state, which is something implementers will overlook easily. So it's easy to say: "eh `Task` and `TaskStruct` are thread-specific, I should not care about concurrency in listeners at all except for contention on output", but that's not the full picture. Being aware of this, your trade off seems more risky to me, so I wanted to make sure you're aware of this as well. Are you still convinced that's the wise choice?

Reply:
> This means that all the listener need to guard their internal state

Yes.

The choice is between us forcing implementors to be safe but inefficient or ask implementors to acquire a lock on each method invocation if they deem it's the safest thing.

But this means we have to pay a cache flush (i.e., acquire a lock) every time we notify an event as opposed to only when we need to actually serialize output to disk/`stderr`.
Maybe this is premature optimization, but the selling point of this thing is "adopt it more or less wherever you have a `for`". If I can make a one time effort to do things efficiently to let every user benefit from this and not be too afraid about performance impact of adoption, I'd be happy to.

What do you think?

Reply:
I think that it depends if you think of this mainly as a mechanism for interfacing with UI or for other uses.

If we're interfacing in real time with the UI I don't think we have to worry about cache lines too much, they're not gonna be the issue. But if we're thinking of using this to log events for profiling it's a totally different story.

And it also depends on the granularity of tasks, which at the moment is not entirely clear to me, and is subject to change on wider adoption of this, and across projects.

I'd love to have a slow-but-always-safe version of the thing, with automatic handling of concurrency, and then a version for grown-ups where you take full responsibility of what you're doing if you're trying to squeeze out performance where it matters. But if we have to choose just one, then ok, let's just do the fast-but-dangerous. We'll just have to take extra care when implementing listeners. I decided to be pedantic because I looked into the WIP listeners and they were broken. If that's something you're aware of let's just go ahead with the fast-but-dangerous thing.

Reply:
> If we're interfacing in real time with the UI

Not real time, we will sample the state every now and then.
Locking every 50ms it's different than locking at every iteration of a loop.

Apart from this, there will be few listeners and many many place where we use `Task`.
Apart from the existing one, we'll just have an additional one for sending status updates to the UI over websockets every now and then. We will not send to the UI every event.

> I'd love to have a slow-but-always-safe version of the thing, with automatic handling of concurrency, and then a version for grown-ups where you take full responsibility

I think this is overkill at this stage. We have 3-4 implementations under our full control. Maybe in a future iteration if we decide to upstream this.

Reply:
Ok, I think we agree on keeping only the fast-but-dangerous version.